server/testnet.go

```diff
@@ -0,0 +1,174 @@
+package server
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	gc "github.com/cosmos/cosmos-sdk/server/config"
+
+	"github.com/cosmos/cosmos-sdk/wire"
+	"github.com/spf13/viper"
+	cfg "github.com/tendermint/tendermint/config"
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+var (
+	nodeDirPrefix = "node-dir-prefix"
+	nValidators   = "v"
+	outputDir     = "o"
+
+	startingIPAddress = "starting-ip-address"
+)
+
+const nodeDirPerm = 0755
+
+// get cmd to initialize all files for tendermint testnet and application
+func TestnetFilesCmd(ctx *Context, cdc *wire.Codec, appInit AppInit) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "testnet",
+		Short: "Initialize files for a Gaiad testnet",
+		Long: `testnet will create "v" number of directories and populate each with
+necessary files (private validator, genesis, config, etc.).
+
+Note, strict routability for addresses is turned off in the config file.
+
+Example:
+
+	gaiad testnet --v 4 --o ./output --starting-ip-address 192.168.10.2
+	`,
+		RunE: func(_ *cobra.Command, _ []string) error {
+			config := ctx.Config
+			err := testnetWithConfig(config, ctx, cdc, appInit)
+			return err
+		},
+	}
+	cmd.Flags().Int(nValidators, 4,
+		"Number of validators to initialize the testnet with")
+	cmd.Flags().String("o", "./mytestnet",
+		"Directory to store initialization data for the testnet")
+	cmd.Flags().String(nodeDirPrefix, "node",
+		"Prefix the directory name for each node with (node results in node0, node1, ...)")
+
+	cmd.Flags().String(startingIPAddress, "192.168.0.1",
+		"Starting IP address (192.168.0.1 results in persistent peers list ID0@192.168.0.1:46656, ID1@192.168.0.2:46656, ...)")
+	return cmd
+}
+
+func testnetWithConfig(config *cfg.Config, ctx *Context, cdc *wire.Codec, appInit AppInit) error {
+
+	// Generate private key, node ID, initial transaction
+	for i := 0; i < viper.GetInt(nValidators); i++ {
+		nodeDirName := fmt.Sprintf("%s%d", viper.GetString(nodeDirPrefix), i)
+		nodeDir := filepath.Join(viper.GetString(outputDir), nodeDirName, "gaiad")
+		clientDir := filepath.Join(viper.GetString(outputDir), nodeDirName, "gaiacli")
+		gentxsDir := filepath.Join(viper.GetString(outputDir), "gentxs")
+		config.SetRoot(nodeDir)
+
+		err := os.MkdirAll(filepath.Join(nodeDir, "config"), nodeDirPerm)
+		if err != nil {
+			_ = os.RemoveAll(viper.GetString(outputDir))
+			return err
+		}
+
+		err = os.MkdirAll(clientDir, nodeDirPerm)
+		if err != nil {
+			_ = os.RemoveAll(viper.GetString(outputDir))
```

Comment:
we should add `viper.GetString(outputDir)` to a variable instead of calling it so many times

Reply:
Ok, I'm on it.